WordPress/openverse-api
# Move media type categories to constants module

## Fixes


Fixes #558 by @dhruvkb

Supersedes #560

## Description


This PR uses the method described by @dhruvkb in #560 to move the media type categories to a constants module. In doing so, it also changes the category serializer field from a `CharField` to a `MultipleChoiceField`. This has some implications on how we convert the arguments into an Elasticsearch query, as we now need to account for the fact that the field's data is a `set` rather than a string that can be `split`.

I'm not sure this is the correct approach currently because we actually _lose_ some useful information in the validation error. Here's an example on `http://localhost:8000/v1/images/?q=wolf&category=sound`.

**Before**

```
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "detail": {
        "category": [
            "Invalid category: sound. Available options: {'illustration', 'photograph', 'digitized_artwork'}"
        ]
    }
}
```

**After**

```
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "detail": {
        "category": [
            "\"sound\" is not a valid choice."
        ]
    }
}

```

I don't know enough about DRF, but perhaps we might be able to alter the error message for category so it does display the acceptable values. I think this will require us to make a custom serializer field.

An _entirely different_ approach we could take is just to move the sets that are defined in the `validate_categories` functions on each serializer and move only that to the constants module. This would prevent us from having to handle the `set` type differently since it would still be a `CharField`, and we're already using the `_validate_enums` function elsewhere.

## Testing Instructions

`just api-test` for unit tests

Use the link above to see the above results, then change the category to `photograph` or similar and see that results get returned.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
